src/foam/nanos/auth/UserAndGroupAuthService.js

```diff
@@ -0,0 +1,609 @@
+        }
+
+        return user;
+      `
+    },
+    {
+      name: 'generateChallenge',
+      documentation: `A challenge is generated from the userID provided. This is
+        saved in a LinkedHashMap with TTL of 5.`,
+      type: 'String',
+      args: [
+        {
+          class: 'Long',
+          name: 'userID'
+        }
+      ],
+      javaThrows: ['foam.nanos.auth.AuthenticationException'],
+      javaCode: `
+        throw new UnsupportedOperationException("Unsupported operation: generateChallenge");
+      `
+    },
+    {
+      name: 'challengedLogin',
+      documentation: `Checks the LinkedHashMap to see if the the challenge
+        supplied is correct and the TTL is still valid.
+
+        How often should we purge this map for challenges that have expired?`,
+      type: 'foam.nanos.auth.User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'userID',
+          type: 'Long'
+        },
+        {
+          name: 'challenge',
+          type: 'String'
+        }
+      ],
+      javaThrows: ['foam.nanos.auth.AuthenticationException'],
+      javaCode: `
+        throw new UnsupportedOperationException("Unsupported operation: challengedLogin");
+      `
+    },
+    {
+      name: 'userAndGroupContext',
+      documentation: `Logs user and sets user group into the current sessions
+        context.`,
+      type: 'foam.nanos.auth.User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'user',
+          type: 'foam.nanos.auth.User'
+        },
+        {
+          name: 'password',
+          type: 'String'
+        }
+      ],
+      javaThrows: ['foam.nanos.auth.AuthenticationException'],
+      javaCode: `
+        if ( user == null ) {
+          throw new AuthenticationException("User not found");
+        }
+
+        // check if user enabled
+        if ( ! user.getEnabled() ) {
+          throw new AuthenticationException("User disabled");
+        }
+
+        // check if user login enabled
+        if ( ! user.getLoginEnabled() ) {
+          throw new AuthenticationException("Login disabled");
+        }
+
+        // check if user group enabled
+        Group group = (Group) groupDAO_.find(user.getGroup());
+        if ( group != null && ! group.getEnabled() ) {
+          throw new AuthenticationException("User group disabled");
+        }
+
+        if ( ! Password.verify(password, user.getPassword()) ) {
+          throw new AuthenticationException("Invalid Password");
+        }
+
+        // Freeze user
+        user = (User) user.fclone();
+        user.freeze();
+
+        Session session = x.get(Session.class);
+        session.setUserId(user.getId());
+        session.setContext(session.getContext().put("user", user));
+
+        return user;
+      `
+    },
+    {
+      name: 'login',
```

Comment:
You can omit the return type and args of methods that you're implementing on the interface. FOAM picks those up for you.